### ISSUE TITLE
When to use map over flatMap

### ISSUE THREAD

**Issue by Herrekt** [nus-cs2030/2021-s2#97]
# Summary

During level 5+ we were to obtain an Optiona<String> at the end representing getGrade( ).

## Description:

When trying to get the grade, I started to flatMap from the Roster, using (x -> x.get(y)) to progress throughout till getGrade( ). However, at the end, I realised only map works for (x -> x.getGrade( )) and not flatMap. I was thinking flatMap and map should be somewhat interchangeable but wasn't sure what was wrong.

## Screenshots
![Capture](https://user-images.githubusercontent.com/80625519/111970843-b4a2e000-8b36-11eb-889c-8036bdbc4b1f.PNG)

**Comment by dumbasssoul** [comment]
the return type of getGrade() is just a String and not Optional<V>, thus there is no need to flatmap it (unlike the return type of get(student) or get(module), which is Optional<V>)

**Comment by kexinnnwang** [comment]
yup you use flatMap if you are expecting an Optional to be returned because you dont want to end up with Optional in Optionals. since getGrade() returns a String, you wont have to worry about this so you can just use map.

**Comment by Onelayer** [comment]
flatMap() is used when the output is an optional as you do not want the output to be an optional inside an optional. In this sense, flatMap() is the same as map(), just that flatMap flattens the output such that there won't be the issue of optionals existing inside an optional. map will be used in cases such as when the output is just a object, so mapping it will return a optional.

**Comment by cf6600** [comment]
When something the mapper function does something to a value of any contained, lets say for an optional, before it gets outputted, it will be wrapped around the optional again, just that this time it contains the mapped value. However, for flatmap, it can be used to remove the optional "contained" and just give you the transformed values which you might want to use in some cases where you want to remove the container.

**Comment by miao-tianjia** [comment]
If your input is already Optional, then you should use a Flatmap so the output will be an Optional instead of an Optional of Optional; if your input is just a String, then using Map is sufficient to output an Optional.

**Comment by karintang** [comment]
Use flatmap when the return type is optional and map if the return type is a string

**Comment by JWulaXia** [comment]
flatmap is when you want to get the content under optional of the optional, which is two layers thus you can flat it, and wrap it under one layer of optional, in this case, get grade() returns a string and you can just wrap it under one layer

**Comment by Martini-Chwang** [comment]
flatmap is used when u wish to maintain the 'Container' of the item inside it, which in this case is to preserve the optional instead of getting an optional in an optional.
